src/packagedcode/conan.py

```diff
@@ -0,0 +1,281 @@
+
+    @classmethod
+    def parse(cls, location):
+        with io.open(location, encoding="utf-8") as loc:
+            conan_recipe = loc.read()
+
+        try:
+            tree = ast.parse(conan_recipe)
+            recipe_class_def = next(
+                (
+                    node
+                    for node in tree.body
+                    if isinstance(node, ast.ClassDef)
+                    and node.bases
+                    and isinstance(node.bases[0], ast.Name)
+                    and node.bases[0].id == "ConanFile"
+                ),
+                None,
+            )
+
+            parser = ConanFileParser()
+            parser.visit(recipe_class_def)
+        except SyntaxError as e:
+            if TRACE:
+                logger_debug(f"Syntax error in conan recipe: {e}")
+            return
+
+        if TRACE:
+            logger_debug(f"ConanFileHandler: parse: package: {parser.to_dict()}")
+
+        dependencies = get_dependencies(parser.requires)
+
+        yield models.PackageData(
+            datasource_id=cls.datasource_id,
+            type=cls.default_package_type,
+            primary_language=cls.default_primary_language,
+            namespace=None,
+            name=parser.name,
+            version=parser.version,
+            description=parser.description,
+            homepage_url=parser.homepage_url,
+            keywords=parser.keywords,
+            extracted_license_statement=parser.license,
+            dependencies=dependencies,
+        )
+
+
+class ConanDataHandler(models.DatafileHandler):
+    datasource_id = "conan_conandata_yml"
+    path_patterns = ("*/conandata.yml",)
+    default_package_type = "conan"
+    default_primary_language = "C++"
+    description = "conan external source"
+    documentation_url = (
+        "https://docs.conan.io/2/tutorial/creating_packages/"
+        "handle_sources_in_packages.html#using-the-conandata-yml-file"
+    )
+
+    @classmethod
+    def parse(cls, location):
+        with io.open(location, encoding="utf-8") as loc:
+            conan_data = loc.read()
+
+        conan_data = saneyaml.load(conan_data)
+        sources = conan_data.get("sources", {})
+
+        for version, source in sources.items():
+            sha256 = source.get("sha256", None)
+            urls = []
+            source_urls = source.get("url")
+            if isinstance(source_urls, str):
+                urls.append(source_urls)
+            elif isinstance(source_urls, list):
+                urls.extend(source_urls)
+
+            for url in urls:
+                yield models.PackageData(
+                    datasource_id=cls.datasource_id,
+                    type=cls.default_package_type,
+                    primary_language=cls.default_primary_language,
+                    namespace=None,
+                    version=version,
+                    download_url=url,
+                    sha256=sha256,
+                )
+
+    @classmethod
+    def assemble(
```

Comment:
Usually we have the assemble function under a base class and then inherit that in all other datafile handlers if these manifests can exist on their own and also together, to cover all the cases, would that also benifit us here?

Reply:
conanfile.py can exist on its own, but conandata.yml should not exist in isolation. We need the custom assembly only in the case of conandata.yml

Reply:
Which brings to light they limits of a single-file-at-a-time approach. Should we start looking outside of the current file? ScanCode is not design for this at all though for now.